### PR TITLE
Rebuild debugger on new Task-based runtime api

### PR DESCRIPTION
**Pull request by vilterp**
Main API is in `Debugger/RuntimeApi.elm`, which uses native module `Native/Debugger/RuntimeApi.js`. All interaction with the module being debugged (time-travel, swapping, new frame notifications, logging, etc) goes through this new API.

- [Design doc explaining the motivation for this API](https://docs.google.com/document/d/1n1rz061cgqHCE9aG4qRInhizV15b34mylwxauiFqCvY/edit)
- [Depends on these changes to core (PR)](https://github.com/elm-lang/core/pull/306)
- [Test plan](https://docs.google.com/document/d/1nCNectlax_0Ms3RywYDl2VTxmirM6gqlHhSMffojezg/edit?usp=sharing)

TODOs before merging:

- bugs
    - [x] debugger shouldn't own the screen, because the virtualdom version it is using to render could be different from the debugee's, where it got the value of `main` from.
    - [ ] mouse events not blocked by sidebar
    - [ ] fix timers
    - [ ] fix socket disconnecting from Haskell side after ~30secs
    - [ ] make pretty printer not crash on `null`
    - [ ] handle differences between values in debuggee's `core` & values debugger's pretty-printer is
expecting (e.g. `RBNode` => `RBNode_elm_special` or whatever it is now)
    - [ ] make sure output ports are not triggered during scrubbing
    - [ ] `getNodeState` in `Debugger.RuntimeApi` has the side effect of causing a re-render. Maybe disconnect renderer in Runtime.js while the session is paused?
    - [ ] node subscriptions are updated on every frame, but should only be updated when the node's value changes. Add a `changedThisFrame` flag or similar to signal graph nodes?
- import/export
    - [x] * fix first frame not showing up
    - [x] nice error messages for invalid import (module name mismatch, etc)
- code cleanup
    - [ ] decompose chained tasks into named things
    - [ ] JS indentation
    - [ ] Elm: 80 chars
    - [ ] make sure Task errors in JS match with Elm type signatures
    - [ ] remove `Debug.log`s
    - [ ] move RuntimeApi into its own repo
    - [ ] use `elm-effects`
    - [ ] get rid of submodules (`elm-websockets` and `elm-html-file`): native review & publish
- design cleanup in separate issue [WIP: UI Questions](https://github.com/elm-lang/elm-reactor/issues/139)
- optimizations
    - [ ] snapshot cache for swapping
    - [ ] scrubbing forward
- API niceties
    - [ ] use 2 modules, so pausing isn’t necessary for getNodeState
    - [ ] take subscription stuff out of `initializeFullscreen`; use other functions for that

### CODE REVIEW

File: frontend/Native/Debugger/RuntimeApi.js [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+						}
+					} else {
+						nodeType = {ctor: 'InternalNode'};
+					}
+					var info = {
+						_: {},
+						name: node.name,
+						nodeType: nodeType,
+						kids: List.fromArray(
+							node.kids ? node.kids.map(function(kid) {return kid.id}) : []
```
Comment (vilterp) [inline frontend/Native/Debugger/RuntimeApi.js]: fixes #136

File: backend/Socket.hs [hunk 2]
```diff
@@ -18,41 +19,21 @@ fileChangeApp :: FilePath -> WS.ServerApp
...
-                case controlMessage of
-                  WS.Ping _ -> listen pingThread
-                  WS.Pong _ -> loop
-                  WS.Close _ _ ->
-                      killThread pingThread >> return ()
+loopForever :: IO ()
+loopForever =
+  do
+    threadDelay (10 * 1000000) -- 10 seconds
+    loopForever
```
Comment (vilterp) [inline backend/Socket.hs]: TODO: find a way to kill this when the socket is closed

**Comment by 0xR** [comment]
I am having the different instance of the same issue as in #136. It would be great if this PR gets review / merged quickly since this apparently contains a fix.

**Comment by 0xR** [comment]
I just looked at the size of the change and may I suggest releasing the fix for #136 in a separate PR? Since this is a complete rewrite it will probably take a while to get reviewed / merged. The fix seems to be a one liner.

File: backend/Socket.hs (hunk 2 above)
Comment (evancz) [inline backend/Socket.hs]: This is a very important TODO item.

**Pull request closed without merging**